File: atcoder/abc/go/abc167/abc167_F/abc167_F.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"sort"
	"strconv"
)

type scanner struct { sc *bufio.Scanner }
func newScanner(input io.Reader) *scanner {
	sc := bufio.NewScanner(input)
	sc.Split(bufio.ScanWords)
	sc.Buffer(make([]byte, 1024), int(1e+9))
	return &scanner{sc}	
}
func (s *scanner) s() string  { s.sc.Scan(); return s.sc.Text() }
func (s *scanner) i() int     { i,e := strconv.Atoi(s.s()); if e != nil {panic(e)}; return i }

var rdr = newScanner(os.Stdin)
const BUFSIZE = 10000000

func min(a,b int) int { if a > b { return b }; return a }
type PI struct { x,y int }

func evalChain(ch []PI) (int,bool) {
	ht := 0
	for _,c := range(ch) {
		if c.x > ht { return 0,false }
		ht += c.y
	}
	return ht,true
}

func main() {
	//f1, _ := os.Create("cpu.prof"); pprof.StartCPUProfile(f1); defer pprof.StopCPUProfile()
	infn := ""
	if infn == "" && len(os.Args) > 1 {	infn = os.Args[1] }
	if infn != "" {	f, e := os.Open(infn); if e != nil { panic(e) }; rdr = newScanner(f) }
	
    // NON-BOILERPLATE STARTS HERE
	N := rdr.i()
	left := make([]PI,0)
	right := make([]PI,0)
	for i:=0;i<N;i++ {
		s := rdr.s()
		low,running := 0,0
		for _,c := range s {
			if c == '(' { running++ } else { running-- }
			low = min(low,running)
		}
		if running >= 0 { 
			right = append(right,PI{-low,running})
		} else { 
			left  = append(left,PI{running-low,-running})
		}
	}
	sort.Slice(left,func(i,j int) bool { return left[i].x < left[j].x })
	sort.Slice(right,func(i,j int) bool { return right[i].x < right[j].x })
	ht1,ok1,ht2,ok2 := 0,false,0,false
	ht1,ok1 = evalChain(left)
	if ok1 { ht2,ok2 = evalChain(right) }
	ans := "No"
	if ok1 && ok2 && ht1==ht2 { ans = "Yes" }
	fmt.Println(ans)
    //fmt.Fprintln(wrtr, ans); wrtr.Flush()
}

File: atcoder/abc/python/abc167/abc167_F.py

import sys
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

def check(up) :
    s = 0
    for (m,d) in up :
        if m > s : return False
        s += d
    return True

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    N = gi(); S = []
    for _ in range(N) : S.append(gs())
    up = []; dn = []
    for s in S :
        d,m = 0,0
        for c in s :
            d += (1 if c == '(' else -1)
            m = min(m,d)
        if d >= 0 : up.append((0-m,d))
        else      : dn.append((d-m,-d))
    up.sort()
    dn.sort()
    ok = True
    if sum(x[1] for x in up) != sum(x[1] for x in dn) : ok = False
    if ok and not check(up) : ok = False
    if ok and not check(dn) : ok = False
    ans = "Yes" if ok else "No"
    print(ans)
    
if __name__ == '__main__' :
    main()
    sys.stdout.flush()

